rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,565 @@
+  // The root slot can never be null, so also exclude null from it
+  [Key in keyof Slots]: ReplaceNullWithUndefined<
+    Exclude<Slots[Key], SlotShorthandValue | (Key extends 'root' ? null : never)>
+  >;
+};
+```
+
+After:
+
+```ts
+type ComponentState<Slots extends SlotPropsRecord> = {
+  [Key in keyof Slots]: SlotComponent<ExtractSlotProps<Slots[Key]>>;
+};
+```
+
+#### Styles hooks
+
+A minor modification in style hooks are required. at the moment we mutate the merged properties provided by `resolveShorthand`
+to include our custom `className` property.
+
+In this case we have some alternatives:
+
+##### Option 1 mutate external properties
+
+Instead of mutating the merged properties, simply mutate the External properties from the slot component:
+
+```diff
+// useAccordionHeader
+export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
+  const styles = useStyles();
+- state.root.className = mergeClasses(
++ state.root.props.className = mergeClasses(
+    accordionHeaderClassNames.root,
+    styles.root,
+    state.inline && styles.rootInline,
+    state.disabled && styles.rootDisabled,
+-   state.root.className,
++   state.root.props.className,
+  );
+
+  return state;
+};
+```
+
+##### Option 2 mutate overrides
+
+By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can be mutated:
+
+```diff
+// useAccordionHeader
+export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
+  const styles = useStyles();
+- state.root.className = mergeClasses(
++ state.overrides.root.className = mergeClasses(
+    accordionHeaderClassNames.root,
+    styles.root,
+    state.inline && styles.rootInline,
+    state.disabled && styles.rootDisabled,
+-   state.root.className,
++   state.root.props.className,
+  );
+
+  return state;
+};
+```
+
+On the render function side, we just need to spread the overrides:
+
+```tsx
+export const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
+  <state.root {...state.overrides.root}>
+    <state.button>
+      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
+      {state.icon && <state.icon />}
+      {state.root.props.children}
+      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
+    </state.button>
+  </state.root>
+);
+```
+
+##### Option 3 stop mutating
+
+Let's just stop mutating, and since we require a layer of custom styling through classNames
+let's just properly use an argument on the render method exclusively for this:
+
+Before:
+
+```diff
+// useAccordionHeader
+-export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
++export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState): AccordionHeaderStyles => {
+  const styles = useStyles();
++ return {
+- state.root.className = mergeClasses(
++   root: mergeClasses(
```

Comment:
You only use `root` in the example, but you would need to do the same for every slot right?

Reply:
Could you please update examples to reflect that?